### PR TITLE
Add Socks Proxy (for Tor/i2pd/Kovri) Support to Wallets

### PR DESCRIPTION
**Pull request by vtnerd**
Allows wallets to connect to daemon over tor/i2pd/kovri. If proxy usage is enabled, the daemon address must be onion/i2p based or use `--daemon-cert-file`. The idea is to prevent root CA shadiness over these networks in addition to the authenticated/unencrypted problems. A root CA can be combined with an onion/i2p address (the former can definitely be signed by a few CAs) by using `--daemon-address https://xyz.onion`. Additionally, `--daemon-address https://...` and `--daemon-cert-file` will work with ipv4 or ICANN hostnames.

Will have to coordinate with @moneromooo-monero for the SSL changes. The first commit (for p2p tor/socks) will be rebased away once its merged.

### CODE REVIEW

**Comment by vtnerd** [comment]
Pushed [some tests and rebased](https://github.com/monero-project/monero/compare/d37ef70..ff04506) against the parent Tor PR.

**Comment by fuwa0529** [comment]
I think the restriction on domain names / CA is a bit too strict, or even irrelevant. It's always possible to ask the socks5 server to resolve the domain on behalf of the user, and assuming the socks5 server is doing the right thing, the user is just as safe as using an IP. I also might want to use a different socks5 server other then tor/i2p.

Is it possible to decouple the commit into smaller ones for educational purpose?

**Comment by vtnerd** [comment]
> I think the restriction on domain names / CA is a bit too strict, or even irrelevant. It's always possible to ask the socks5 server to resolve the domain on behalf of the user, and assuming the socks5 server is doing the right thing, the user is just as safe as using an IP. I also might want to use a different socks5 server other then tor/i2p.

The socks server is always doing the domain resolution. If a user wants to use a socks proxy in some other context, they can download the certificate of the CA for that site and specify with `--daemon-cert-file`. This is a bit restrictive, but allowing all root CAs on outbound proxy connections seemed risky without some nudging.

> Is it possible to decouple the commit into smaller ones for educational purpose?

There are already two commits - one from the PR for p2p+tor, and a second for socks rpc. Did you want this broken down further?

**Comment by fuwa0529** [comment]
Thanks for the explanation @vtnerd. I see now that the default is to not trust root CA, then this seems to be the best practice.

Yes, I was hoping the commits can be broken down further more, if possible?

**Comment by jtgrassie** [comment]
@fuwa0529 
> Yes, I was hoping the commits can be broken down further more, if possible?

Why? The 2 commits are very logical units of work.

File: contrib/epee/src/net_helper.cpp [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+				m_io_service.run_one();
+
+				if (connection.is_ready())
+					break;
+			}
+
+			m_ssl_stream.emplace(m_io_service);
+			m_ssl_stream->socket.next_layer() = connection.get();
+			m_deadline.cancel();
+			m_deadline.expires_at(std::chrono::steady_clock::time_point::max());
```
Comment (moneromooo-monero) [inline contrib/epee/src/net_helper.cpp]: Is this scheduling an event for essentially never ?

**Comment by fluffypony** [comment]
@vtnerd as soon as #4852 is rebased and merged I'll let you know, and this can be rebased and merged!

**Comment by MaxXor** [comment]
It is merged.

**Comment by fluffypony** [comment]
@vtnerd please rebase

**Comment by fluffypony** [comment]
@vtnerd still waiting on a rebase :)

File: ANONYMITY_NETWORKS.md [hunk 2]
```diff
@@ -74,6 +79,36 @@ forwarded to `monerod` localhost port 30000.
 These addresses will be shared with outgoing peers, over the same network type,
 otherwise the peer will not be notified of the peer address by the proxy.
 
+### Wallet RPC
+
+An anonymity network can be configured to forward incoming connections to a
+`monerod` RPC port - which is independent from the configuration for incoming
+P2P anonymity connections. The anonymity network (tor, i2pd, kovri, etc.) is
```
Comment (jtgrassie) [inline ANONYMITY_NETWORKS.md]: Minor request, can we change "(tor, i2pd, kovri, etc.)" to simply "tor/i2p"? This keeps implementation names out.

File: ANONYMITY_NETWORKS.md [hunk 3]
```diff
@@ -74,6 +79,36 @@ forwarded to `monerod` localhost port 30000.
...
+[configured in the same manner](#configuration), except the localhost port
+must be the RPC port (typically 18081 for mainnet) instead of the p2p port:
+
+> HiddenServiceDir /var/lib/tor/data/monero
+> HiddenServicePort 18081 127.0.0.1:18081
+
+Then the wallet will be configured to use a tor/i2pd/kovri proxy and onion/i2p
```
Comment (jtgrassie) [inline ANONYMITY_NETWORKS.md]: Same as above comment.

File: ANONYMITY_NETWORKS.md (hunk 2 above)
Comment (vtnerd) [reply ANONYMITY_NETWORKS.md]: Will update shortly.

**Comment by vtnerd** [comment]
> @vtnerd still waiting on a rebase :)

Yeah, I got caught working on the SSL PR. This has reverted prior changes for SSL, and is rebased to mainline. @moneromooo-monero may want to review shortly since it technically has changed, although  the "new" changes are primarily just reverting changes that were previously made.

File: ANONYMITY_NETWORKS.md (hunk 2 above)
Comment (jtgrassie) [reply ANONYMITY_NETWORKS.md]: Thanks

**Comment by vtnerd** [comment]
Ok, finally rebased entirely (hopefully) with @jtgrassie changes.

**Pull request merged**